core/src/main/java/org/elasticsearch/action/support/single/instance/TransportInstanceSingleOperationAction.java

```diff
@@ -214,6 +216,15 @@ public void handleException(TransportException exp) {
         void retry(final @Nullable Throwable failure) {
             if (observer.isTimedOut()) {
                 // we running as a last attempt after a timeout has happened. don't retry
+                Throwable listenFailure = failure;
+                if (listenFailure == null) {
+                    if (shardIt == null) {
+                        listenFailure = new ElasticsearchTimeoutException("Timeout waiting for [" + request.timeout() + "], request: " + request.toString());
```

Comment:
can we use the new fancy string formatting support in our Exceptions? Also, I'm not sure request.toString makes sense (I know it was how it was done), not all requests implement it  (I think!) and we will just have an ugly pointer. Maybe use the action name instead?